File: Leetcode/05_group_anagram.py
# https://leetcode.com/problems/group-anagrams/
from typing import List


def groupAnagrams(strs: List[str]) -> List[List[str]]:
    dict = {}
    for item in strs:
        # ? list는 dictionary의 key가 될 수 없다. (immutable type만 key로 사용 가능)
        key = tuple(sorted(item))
        if key in dict:
            dict[key].append(item)
        else:
            dict[key] = []
            dict[key].append(item)

    return [dict[items] for items in dict]


groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])

File: Leetcode/05_group_anagram.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  /**
   *  정렬 한 문장을 해쉬테이블의 키로 사용하고
   *  원본 문장을 배열에 추가
   *
   */
  const result = strs.reduce((p, c) => {
    let sortedWord = c.split('').sort().join('');
    if (p[sortedWord]) {
      p[sortedWord].push(c);
    } else {
      p[sortedWord] = [c];
    }
    return p;
  }, {});

  return Object.keys(result).map((v) => [...result[v]]);
};

console.log(groupAnagrams(['eat', 'tea', 'tan', 'ate', 'nat', 'bat']));
